lib/new_relic/agent/instrumentation/queue_time.rb

```diff
@@ -27,7 +27,23 @@ module QueueTime
 
         module_function
 
-        def parse_frontend_timestamp(headers, now=Time.now)
+        def frontend_timestamp(headers, now=Time.now)
+          parsed_timestamp = parse_frontend_timestamp(headers, now)
+
+          # When enabled, this will return an adjusted timestamp
+          # with request buffer time subtracted.  This is the time
+          # spent waiting for the HTTP request body to be received.
+          if NewRelic::Agent.config[:'queue_time_subtract_buffering']
+            # Currently, only Puma is supported
+            request_buffering_in_seconds = headers.fetch('puma.request_body_wait', 0).to_f / 1000
```

Comment:
How hard to extend this to other rack providers (webrick, thin, unicorn, etc.)?

Reply:
Let me research this and I'll let you know what I find 👍

Reply:
Hi, @mwlang, hope you're well.  I was hoping other web servers produced a similar header like `puma.request_body_wait` but didn't find anything like this for other commonly used web servers.

Nate Berkopec has a great chapter `Webservers and I/O models` in his book `The Complete Guide to Rails Performance` that covers different web server models.  In it, he describes which web servers offer a model that download larger requests before passing them to a worker to be run application code.

I went through each of these servers and tested a 15kB PUT on a simulated slow 3G connection.

### Webrick
Webrick doesn't buffer large requests.  I also didn't see any passed headers that log the buffering time.

### Thin
Thin can buffer large requests but I didn't see any passed headers that log buffering time.

### Unicorn
I didn't see any passed headers that log buffering time.  It looks like it's recommended to use with nginx to buffer requests.  I'm an nginx beginner but wondering if it's possible set a header when a request has finish buffering so we could calculate here what the buffering time is.

### Phusion Passenger
Phusion Passenger also uses nginx.  I didn't see any headers logging the request buffer time.  Also wondering if it would be possible to set a header in nginx when a request has completed buffering so we could then calculate the buffering time here.